ding/policy/decision_transformer.py

```diff
@@ -0,0 +1,409 @@
+        context_len=20,
+        n_blocks=3,
+        embed_dim=128,
+        n_heads=1,
+        dropout_p=0.1,
+        learn=dict(
+            # (bool) Whether to use multi gpu
+            multi_gpu=False,
+            # How many updates(iterations) to train after collector's one collection.
+            # Bigger "update_per_collect" means bigger off-policy.
+            # collect data -> update policy-> collect data -> ...
+            update_per_collect=1,
+            # batch_size=64,
+            learning_rate=1e-4,
+            # ==============================================================
+            # The following configs are algorithm-specific
+            # ==============================================================
+        ),
+        # collect_mode config
+        collect=dict(),
+        eval=dict(),
+        # other config
+        other=dict(),
+    )
+
+    def _init_learn(self) -> None:
+        r"""
+            Overview:
+                Learn mode init method. Called by ``self.__init__``.
+                Init the optimizer, algorithm config, main and target models.
+            """
+
+        self.stop_value = self._cfg.stop_value
+        self.env_name = self._cfg.env_name
+        dataset = self._cfg.dataset  # medium / medium-replay / medium-expert
+        # rtg_scale: scale of `return to go`
+        # rtg_target: max target of `return to go`
+        # Our goal is normalize `return to go` to (0, 1), which will favour the covergence. As a result, we usually set rtg_scale == rtg_target.
+        self.rtg_scale = self._cfg.rtg_target  # normalize returns to go
+        self.rtg_target = self._cfg.rtg_target  # max target reward_to_go
+        self.max_eval_ep_len = self._cfg.max_eval_ep_len  # max len of one episode
+        self.num_eval_ep = self._cfg.num_eval_ep  # num of evaluation episodes
+
+        lr = self._cfg.learn.learning_rate  # learning rate
+        wt_decay = self._cfg.wt_decay  # weight decay
+        warmup_steps = self._cfg.warmup_steps  # warmup steps for lr scheduler
+
+        # total updates = max_train_iters x num_updates_per_iter
+        max_train_iters = self._cfg.max_train_iters
+        self.num_updates_per_iter = self._cfg.num_updates_per_iter
+
+        self.context_len = self._cfg.context_len  # K in decision transformer
+        n_blocks = self._cfg.n_blocks  # num of transformer blocks
+        embed_dim = self._cfg.embed_dim  # embedding (hidden) dim of transformer
+        n_heads = self._cfg.n_heads  # num of transformer heads
+        dropout_p = self._cfg.dropout_p  # dropout probability
+
+        # # load data from this file
+        # dataset_path = f'{self._cfg.dataset_dir}/{env_d4rl_name}.pkl'
+
+        # saves model and csv in this directory
+        self.log_dir = self._cfg.log_dir
+        if not os.path.exists(self.log_dir):
+            os.makedirs(self.log_dir)
+
+        # training and evaluation device
+        self.device = torch.device(self._device)
```

Comment:
policy has already had `self._device` attribute

Reply:
Yes, but `self._device` is a variable with dtype `str` and here needs a variable with dtype `torch.device`